File: polyglot/ruby/otel_semantic_parser.rb
require 'json'
require 'time'

module AgentLog
  # ============== DATA MODELS ==============

  # Represents parsed semantic attributes from an OTel span
  SpanAttributes = Struct.new(
    :operation_name,       # e.g., "completion", "chat"
    :system,               # e.g., "gpt-4o", "claude-3-5-sonnet"
    :model_id,             # normalized model identifier
    :input_text,           # concatenated input prompts
    :output_text,          # concatenated completions
    :input_tokens,         # integer or nil
    :output_tokens,        # integer or nil
    :total_tokens,         # integer or nil
    :response_id,          # unique response identifier
    :latency_ms,           # total latency in milliseconds
    :tool_calls,           # array of tool/function calls
    :metadata              # any additional context
  )

  # ============== CONFIGURATION ==============

  GENAI_ATTR_PREFIX = 'gen_ai.'
  
  OPERATIONS = {
    'completion' => 'Completion',
    'chat' => 'Chat',
    'embedding' => 'Embedding',
    'classification' => 'Classification',
    'summarization' => 'Summarization',
    'translation' => 'Translation',
    'transcription' => 'Transcription'
  }

  # ============== PARSER IMPLEMENTATION ==============

  class SemanticParser
    def self.parse(span_attributes)
      new.parse(span_attributes)
    end

    def initialize
      @input_parts = []
      @output_parts = []
      @tool_calls = []
    end

    def parse(raw_attrs)
      # Extract operation name with default fallback
      operation_name = extract_operation(raw_attrs)
      
      # Extract system/provider info
      system = extract_system(raw_attrs)
      
      # Extract model identifier and normalize it
      model_id = extract_model(raw_attrs)
      
      # Extract token usage metrics
      input_tokens = extract_token_count('input', raw_attrs)
      output_tokens = extract_token_count('output', raw_attrs)
      total_tokens = extract_token_count('total', raw_attrs)
      
      # Extract response identifier if present
      response_id = extract_response_id(raw_attrs)
      
      # Extract latency if available
      latency_ms = extract_latency(raw_attrs)
      
      # Build input/output text from message arrays or single prompt/response
      @input_parts.concat(extract_inputs(raw_attrs))
      @output_parts.concat(extract_outputs(raw_attrs))
      
      # Parse tool/function calls
      @tool_calls.concat(extract_tool_calls(raw_attrs))
      
      # Build final structured result
      SpanAttributes.new(
        operation_name: operation_name,
        system: system,
        model_id: model_id,
        input_text: build_combined_text(@input_parts),
        output_text: build_combined_text(@output_parts),
        input_tokens: input_tokens,
        output_tokens: output_tokens,
        total_tokens: total_tokens,
        response_id: response_id,
        latency_ms: latency_ms,
        tool_calls: @tool_calls,
        metadata: extract_metadata(raw_attrs)
      )
    end

    private

    def extract_operation(attrs)
      operation = attrs['gen_ai.operation.name'] || 'unknown'
      
      # Normalize to title case with proper formatting
      normalized = OPERATIONS[operation] || operation.capitalize
      
      # Handle compound operations (e.g., "chat.completion")
      if compound_op = normalize_compound_operation(operation)
        return "#{compound_op}::#{normalized}"
      end
      
      normalized
    end

    def extract_system(attrs)
      system = attrs['gen_ai.system'] || 
               attrs['attributes.gen_ai.system'] ||
               'unknown'
      
      # Normalize common provider prefixes
      case system.downcase
      when /claude/ then "Anthropic::#{system}"
      when /gemini|google/ then "Google::#{system}"
      when /dall-e|vision/ then "OpenAI Vision::#{system}"
      else system
      end
    end

    def extract_model(attrs)
      # Priority: explicit model field, then request.model, then infer from system
      if attrs['gen_ai.request.model']
        return normalize_model_name(attrs['gen_ai.request.model'])
      end
      
      # Try to infer from common patterns
      inferred = infer_model_from_context(attrs)
      
      inferred || 'unknown'
    end

    def extract_token_count(type, attrs)
      key = "attributes.gen_ai.usage.#{type}_tokens"
      value = attrs[key] || attrs["gen_ai.usage.#{type}_tokens"]
      
      return Integer(value) if value
      
      # Calculate from total_tokens and one known type
      if attrs['gen_ai.usage.total_tokens'] && 
         (attrs['gen_ai.usage.input_tokens'].nil? || attrs['gen_ai.usage.output_tokens'].nil?)
        total = Integer(attrs['gen_ai.usage.total_tokens'])
        input = attrs['gen_ai.usage.input_tokens']
        output = attrs['gen_ai.usage.output_tokens']
        
        if input && !output.nil?
          return total - (Integer(input) || 0)
        elsif output && !input.nil?
          return total - (Integer(output) || 0)
        end
      end
      
      nil
    end

    def extract_response_id(attrs)
      attrs['gen_ai.response.id'] || 
      attrs['attributes.gen_ai.response.id'] ||
      attrs['response_id'] ||
      'unknown'
    end

    def extract_latency(attrs)
      # Try multiple possible latency fields
      candidates = [
        'attributes.latency',
        'latency_ms',
        'duration_ms',
        'attributes.duration'
      ]
      
      candidates.each do |candidate|
        value = attrs[candidate]
        return Integer(value) if value && !value.to_s.empty?
      end
      
      nil
    end

    def extract_inputs(attrs)
      # Handle message arrays (chat format)
      messages = attrs['attributes.gen_ai.messages'] || 
                 attrs['messages'] || []
      
      return messages.map { |m| m.is_a?(Hash) ? m['content'] : m.to_s } if messages
      
      # Single prompt/response pair
      if attrs['prompt'] && !attrs['response'].nil?
        [attrs['prompt'], attrs['response']]
      elsif attrs['input']
        [attrs['input']]
      else
        []
      end
    end

    def extract_outputs(attrs)
      # Handle message arrays (chat format)
      messages = attrs['attributes.gen_ai.messages'] || 
                 attrs['messages'] || []
      
      return messages.map { |m| m.is_a?(Hash) ? m['content'] : m.to_s } if messages
      
      # Single response
      [attrs['response'] || attrs['output']]
    end

    def build_combined_text(parts)
      parts.compact.join("\n\n---\n\n")
    end

    def extract_tool_calls(attrs)
      tool_fields = [
        'attributes.gen_ai.tool_calls',
        'tool_calls',
        'function_calls'
      ]
      
      return [] if attrs['gen_ai.operation.name'] == 'embedding' || 
                   attrs['gen_ai.operation.name'] == 'completion'
      
      tools = tool_fields.map { |f| attrs[f] }.compact
      
      # Parse tool call objects into structured format
      parsed_tools = tools.flat_map do |tool_data|
        next [] unless tool_data.is_a?(Array)
        
        tool_data.map do |t|
          {
            name: t['name'] || t['function']['name'],
            arguments: t['arguments'] || t['function']['arguments'],
            id: t['id'] || t['call_id'] || 'unknown'
          }
        end
      end
      
      parsed_tools.flatten(1)
    end

    def extract_metadata(attrs)
      # Extract any custom metadata fields
      {
        version: attrs['attributes.version'] || ENV['AGENTLOG_VERSION'],
        timestamp: Time.now.iso8601,
        span_id: attrs['span_id'] || 'unknown',
        trace_id: attrs['trace_id'] || 'unknown'
      }
    end

    def normalize_model_name(name)
      # Normalize common model name patterns
      normalized = name.to_s.strip
      
      # Handle version suffixes
      if match = normalized.match(/(gpt-|claude-|gemini-)([^0-9]+)?(\d+)?/)
        base, prefix, version = match.captures
        "#{prefix}#{version}"
      else
        normalized
      end
    rescue
      name.to_s.strip
    end

    def infer_model_from_context(attrs)
      # Infer model from system/provider hints
      case attrs['gen_ai.system']&.downcase
      when /claude/ then "claude-3-5-sonnet"
      when /gemini|google.*vision/ then "gemini-1.5-flash"
      when /dall-e/ then "gpt-4o-dalle"
      else nil
      end
    end

    def normalize_compound_operation(operation)
      # Split compound operations like "chat.completion"
      parts = operation.split('.')
      
      return parts.last if parts.length > 1
      
      operation
    end
  end

  # ============== CONVENIENCE METHODS ==============

  class << self
    def parse_span(span_attrs)
      SemanticParser.parse(span_attrs)
    end

    def format_attributes(attrs, pretty: true)
      result = attrs.to_h.transform_values do |v|
        v.is_a?(Time) ? v.iso8601 : (v.is_a?(Integer) ? v : v.to_s)
      end
      
      pretty ? JSON.pretty_generate(result) : JSON.generate(result)
    end

    def create_sample_span(attrs = {})
      default_attrs = {
        'gen_ai.operation.name' => 'chat',
        'gen_ai.system' => 'gpt-4o',
        'gen_ai.request.model' => 'gpt-4o-2024-11-20',
        'gen_ai.usage.input_tokens' => 156,
        'gen_ai.usage.output_tokens' => 89,
        'gen_ai.usage.total_tokens' => 245,
        'gen_ai.response.id' => 'resp_abc123',
        'attributes.latency' => 1250,
        'messages' => [
          { role: 'user', content: 'Hello, how are you?' },
          { role: 'assistant', content: 'I am doing well!' }
        ],
        'tool_calls' => [
          { id: 'call_1', name: 'search_web', arguments: { query: 'weather' } },
          { id: 'call_2', name: 'get_weather', arguments: { location: 'SF' } }
        ]
      }.merge(attrs)

      SemanticParser.parse(default_attrs)
    end
  end

  # ============== DEMO / ENTRY POINT ==============

  if __FILE__ == $0
    puts "=== AgentLog OTel Semantic Parser Demo ==="
    puts ""

    # Sample 1: Basic chat completion with messages
    sample_1 = {
      'gen_ai.operation.name' => 'chat',
      'gen_ai.system' => 'gpt-4o',
      'gen_ai.request.model' => 'gpt-4o-2024-05-13',
      'gen_ai.usage.input_tokens' => 98,
      'gen_ai.usage.output_tokens' => 47,
      'gen_ai.response.id' => 'resp_demo_001',
      'attributes.latency' => 892,
      'messages' => [
        { role: 'user', content: 'What is the capital of France?' },
        { role: 'assistant', content: 'The capital of France is Paris.' }
      ]
    }

    puts "Sample 1 - Basic Chat:"
    attrs_1 = SemanticParser.parse(sample_1)
    puts "  Operation: #{attrs_1.operation_name}"
    puts "  Model:     #{attrs_1.model_id}"
    puts "  Tokens:    I=#{attrs_1.input_tokens}, O=#{attrs_1.output_tokens}, T=#{attrs_1.total_tokens}"
    puts "  Latency:   #{attrs_1.latency_ms}ms"
    puts "  Response:  #{attrs_1.response_id}"
    puts ""

    # Sample 2: With tool calls
    sample_2 = {
      'gen_ai.operation.name' => 'chat',
      'gen_ai.system' => 'claude-3.5-sonnet',
      'messages' => [
        { role: 'user', content: 'Search for the latest news about AI.' },
        { role: 'assistant', content: '', tool_calls: [
          { id: 'call_123', name: 'search_web', arguments: { query: 'AI news 2024' } }
        ]},
        { role: 'tool', content: 'Found 5 articles...', tool_call_id: 'call_123' },
        { role: 'assistant', content: 'Here are the top results...' }
      ],
      'gen_ai.usage.input_tokens' => 203,
      'gen_ai.usage.output_tokens' => 147,
      'attributes.latency' => 2156,
      'gen_ai.response.id' => 'resp_demo_002'
    }

    puts "Sample 2 - With Tool Calls:"
    attrs_2 = SemanticParser.parse(sample_2)
    puts "  Operation: #{attrs_2.operation_name}"
    puts "  System:    #{attrs_2.system}"
    puts "  Model:     #{attrs_2.model_id}"
    puts "  Tokens:    I=#{attrs_2.input_tokens}, O=#{attrs_2.output_tokens}"
    puts "  Latency:   #{attrs_2.latency_ms}ms"
    if attrs_2.tool_calls.any?
      puts "  Tools:     #{attrs_2.tool_calls.size} calls detected"
      attrs_2.tool_calls.each_with_index do |tc, i|
        puts "    [#{i+1}] #{tc[:name]} - #{tc[:arguments].to_s}"
      end
    else
      puts "  Tools:     No tool calls found"
    end
    puts ""

    # Sample 3: Embedding operation (no messages)
    sample_3 = {
      'gen_ai.operation.name' => 'embedding',
      'gen_ai.system' => 'text-embedding-3-small',
      'gen_ai.request.model' => 'text-embedding-3-small',
      'gen_ai.usage.input_tokens' => 12,
      'gen_ai.usage.output_tokens' => 768,
      'attributes.latency' => 45,
      'input': 'Hello world test embedding',
      'response_id': 'embed_resp_003'
    }

    puts "Sample 3 - Embedding:"
    attrs_3 = SemanticParser.parse(sample_3)
    puts "  Operation: #{attrs_3.operation_name}"
    puts "  Model:     #{attrs_3.model_id}"
    puts "  Tokens:    I=#{attrs_3.input_tokens}, O=#{attrs_3.output_tokens}"
    puts "  Latency:   #{attrs_3.latency_ms}ms"
    puts "  Input:     '#{attrs_3.input_text[0..50]}...'" if attrs_3.input_text.length > 50
    puts ""

    # Summary statistics
    puts "=== Summary Statistics ==="
    total_tokens = [attrs_1, attrs_2, attrs_3].sum { |a| a.total_tokens || 0 }
    total_latency = [attrs_1, attrs_2, attrs_3].sum {

File: polyglot/python/otel_semantic_parser.py
"""
polyglot/python/otel_semantic_parser.py

OTel Semantic Parser - Agentic workflow replay & audit utility.

Parses OpenTelemetry GenAI semantic attributes from raw log strings into
structured Python objects for querying and analysis.

Supports:
- Stringified JSON (most common in logs)
- Native Python types (dict, list, int, float, bool, str)
- Mixed/ambiguous formats with auto-detection
"""

import json
from typing import Any, Dict, List, Optional, Union


class SemanticValueError(Exception):
    """Raised when semantic parsing fails."""
    pass


def _detect_type(value: Any) -> type:
    """Detect the actual Python type of a value."""
    if isinstance(value, (dict, list)):
        return type(value)
    if isinstance(value, str):
        # Check for JSON stringification
        try:
            parsed = json.loads(value)
            return _detect_type(parsed)
        except (json.JSONDecodeError, TypeError):
            pass
        return str
    if isinstance(value, (int, float, bool)):
        return type(value)
    return type(value)


def _parse_json_string(s: str) -> Any:
    """Parse a string that may be JSON or plain text."""
    s = s.strip()
    if not s.startswith('"') and not s.startswith('['):
        # Not JSON, treat as plain string
        return s
    
    try:
        parsed = json.loads(s)
        return parsed
    except (json.JSONDecodeError, TypeError):
        return s


def _normalize_value(value: Any) -> Any:
    """Normalize a value to its canonical Python representation."""
    if isinstance(value, str):
        # Try JSON parsing first for strings that look like structured data
        try:
            parsed = json.loads(value)
            if not isinstance(parsed, (str, int, float, bool)):
                return parsed
        except (json.JSONDecodeError, TypeError):
            pass
    
    # Recursively normalize nested structures
    if isinstance(value, dict):
        return {k: _normalize_value(v) for k, v in value.items()}
    if isinstance(value, list):
        return [_normalize_value(item) for item in value]
    
    return value


def parse_semantic_attributes(
    raw_attrs: Union[str, Dict[str, Any]],
    strict_mode: bool = False
) -> Dict[str, Any]:
    """
    Parse OpenTelemetry semantic attributes into structured data.

    Args:
        raw_attrs: Raw attribute string or dict (e.g., from log line)
        strict_mode: If True, raise SemanticValueError on parse failures

    Returns:
        Normalized dictionary of parsed values

    Examples:
        >>> # Stringified JSON
        >>> attrs = '{"model_name": "gpt-4", "response_tokens": 128}'
        >>> result = parse_semantic_attributes(attrs)
        >>> print(result['model_name'])  # 'gpt-4'

        >>> # Mixed types
        >>> attrs = '{"prompt_id": "abc-123", "is_streaming": true, "tokens": [10, 20]}'
        >>> result = parse_semantic_attributes(attrs)
    """
    if isinstance(raw_attrs, dict):
        return _normalize_value(raw_attrs)

    # Try JSON parsing first
    try:
        parsed = json.loads(raw_attrs)
        if not isinstance(parsed, (str, int, float, bool)):
            return _normalize_value(parsed)
    except (json.JSONDecodeError, TypeError):
        pass

    # Fallback to string
    result = str(raw_attrs).strip()
    
    if strict_mode and len(result) > 0:
        SemanticValueError(f"Failed to parse semantic attributes: {raw_attrs[:100]}")

    return result


def extract_genai_fields(
    parsed_data: Dict[str, Any],
    field_mapping: Optional[Dict[str, str]] = None
) -> Dict[str, Any]:
    """
    Extract GenAI-specific fields from parsed semantic data.

    Standard OTel GenAI attribute names per spec:
    - model_name: The LLM model identifier
    - prompt_id: Unique prompt identifier (UUID, trace ID, etc.)
    - response_tokens: Number of tokens in response
    - input_tokens: Number of tokens in user input
    - total_tokens: Combined token count
    - is_streaming: Boolean flag for streaming responses
    - finish_reason: Why generation stopped

    Args:
        parsed_data: Output from parse_semantic_attributes()
        field_mapping: Optional custom mapping (raw_name -> canonical name)

    Returns:
        Dictionary with extracted and normalized GenAI fields
    """
    # Standard OTel GenAI field mappings
    STANDARD_FIELDS = {
        'model_name': ['model_name', 'llm.model', 'model'],
        'prompt_id': ['prompt_id', 'input.id', 'traceId', 'spanId'],
        'response_tokens': ['response_tokens', 'output.tokens', 'tokens.out'],
        'input_tokens': ['input_tokens', 'input.tokens', 'tokens.in'],
        'total_tokens': ['total_tokens', 'tokens.total'],
        'is_streaming': ['is_streaming', 'streaming', 'mode.stream'],
        'finish_reason': ['finish_reason', 'stop_reason', 'completion.reason'],
    }

    result: Dict[str, Any] = {}

    for canonical_name, aliases in STANDARD_FIELDS.items():
        # Search through all nested levels for matching keys
        found_value = _find_nested_key(parsed_data, aliases)
        
        if found_value is not None:
            result[canonical_name] = found_value

    return result


def _find_nested_key(
    data: Any, 
    target_aliases: List[str],
    path: Optional[List[str]] = None
) -> Optional[Any]:
    """Recursively search for a key in nested structures."""
    if isinstance(data, dict):
        # Check current level
        for alias in target_aliases:
            if alias in data:
                return data[alias]
        
        # Recurse into values
        for k, v in data.items():
            found = _find_nested_key(v, target_aliases, path + [k])
            if found is not None:
                return found
    
    elif isinstance(data, list):
        for item in data:
            found = _find_nested_key(item, target_aliases, path)
            if found is not None:
                return found

    return None


def build_audit_record(
    parsed_data: Dict[str, Any],
    timestamp: Optional[str] = None,
    span_id: Optional[str] = None,
) -> Dict[str, Any]:
    """
    Build a structured audit record for replay/analysis.

    Args:
        parsed_data: Output from parse_semantic_attributes()
        timestamp: ISO 8601 timestamp (defaults to 'now')
        span_id: OTel span ID if available

    Returns:
        Complete audit record ready for storage or replay
    """
    genai_fields = extract_genai_fields(parsed_data)

    return {
        '_meta': {
            'timestamp': timestamp,
            'span_id': span_id,
            'source_format': _detect_type(parsed_data).__name__,
        },
        'genai': genai_fields,
        'raw_attributes': parsed_data,
    }


def replay_workflow(
    audit_records: List[Dict[str, Any]],
    field_mapping: Optional[Dict[str, str]] = None
) -> Dict[str, Any]:
    """
    Reconstruct a workflow from multiple audit records.

    Args:
        audit_records: List of records from build_audit_record()
        field_mapping: Custom field mapping for all records

    Returns:
        Workflow reconstruction with timeline and context
    """
    # Sort by timestamp if available
    sorted_records = sorted(
        audit_records,
        key=lambda r: r.get('_meta', {}).get('timestamp') or '',
    )

    workflow = {
        'records': len(sorted_records),
        'timeline': [],
        'context': {},
    }

    for record in sorted_records:
        timeline_entry = {
            'timestamp': record['_meta'].get('timestamp'),
            'span_id': record['_meta'].get('span_id'),
            'model': record['genai'].get('model_name'),
            'tokens_in': record['genai'].get('input_tokens', 0),
            'tokens_out': record['genai'].get('response_tokens', 0),
        }

        # Extract additional context from raw attributes
        timeline_entry['context'] = {k: v for k, v in record['raw_attributes'].items() 
                                   if not k.startswith('_')}

        workflow['timeline'].append(timeline_entry)

    return workflow


# ============================================
# DEMO / ENTRY POINT
# ============================================

if __name__ == '__main__':
    # Sample raw log data (as it might appear in a real system)
    SAMPLE_LOGS = [
        """{"model_name": "gpt-4o", "prompt_id": "uuid-1234-5678", 
         "input_tokens": 150, "response_tokens": 320, 
         "total_tokens": 470, "is_streaming": false,
         "finish_reason": "user_content_filter"}""",
        
        """{"model_name": "claude-3.5", "prompt_id": "uuid-9abc-def1", 
         "input_tokens": 89, "response_tokens": 201, 
         "total_tokens": 290, "is_streaming": true,
         "finish_reason": "max_tokens"}""",
    ]

    print("=" * 60)
    print("OTEL SEMANTIC PARSER - DEMO")
    print("=" * 60)

    # Parse each log entry
    for i, raw_log in enumerate(SAMPLE_LOGS):
        print(f"\n--- Log Entry {i + 1} ---")
        
        # Step 1: Parse the raw attributes
        parsed = parse_semantic_attributes(raw_log)
        print(f"Parsed type: {_detect_type(parsed).__name__}")

        # Step 2: Extract GenAI fields
        genai = extract_genai_fields(parsed)
        print("Extracted GenAI fields:")
        for key, value in sorted(genai.items()):
            print(f"    {key}: {value}")

        # Step 3: Build audit record
        audit = build_audit_record(
            parsed, 
            timestamp="2024-01-15T10:30:00Z",
            span_id=f"span-{i+1}"
        )
        print(f"\nAudit record created with {len(audit)} top-level keys")

    # Step 4: Reconstruct workflow timeline
    print("\n" + "=" * 60)
    print("WORKFLOW RECONSTRUCTION")
    print("=" * 60)

    workflow = replay_workflow([audit for audit in [build_audit_record(
        parse_semantic_attributes(log), 
        extract_genai_fields(parse_semantic_attributes(log))
    ) for log in SAMPLE_LOGS]])

    print(f"Total records: {workflow['records']}")
    print("\nTimeline:")
    for entry in workflow['timeline']:
        print(f"  [{entry['timestamp']}] Model: {entry.get('model') or 'unknown'}, "
              f"Tokens: {entry['tokens_in']} in / {entry['tokens_out']} out")

    # Step 5: Quick summary statistics
    print("\n" + "=" * 60)
    print("SUMMARY STATISTICS")
    print("=" * 60)

    models = set()
    total_tokens = 0
    
    for record in [audit for audit in [build_audit_record(
        parse_semantic_attributes(log), 
        extract_genai_fields(parse_semantic_attributes(log))
    ) for log in SAMPLE_LOGS]]:
        model = record['genai'].get('model_name')
        if model:
            models.add(model)
        total_tokens += record['genai'].get('total_tokens', 0)

    print(f"Unique models: {len(models)}")
    for m in sorted(models):
        print(f"  - {m}")
    print(f"Total tokens processed: {total_tokens:,}")